File: backend/internal/telegram/bot.go
package telegram

import (
	"context"
	"fmt"
	"log/slog"
	"strings"

	"go-service-template/internal/models"

	tgbot "github.com/go-telegram/bot"
	tgmodels "github.com/go-telegram/bot/models"
	"github.com/google/uuid"
)

// botUserRepo is the minimal interface the bot needs for account linking and lookups.
type botUserRepo interface {
	SetTelegramID(ctx context.Context, userID uuid.UUID, telegramID int64) (*models.User, error)
	IsTelegramIDTaken(ctx context.Context, telegramID int64, excludeUserID uuid.UUID) (bool, error)
	GetTelegramID(ctx context.Context, userID uuid.UUID) (*int64, error)
	GetByTelegramID(ctx context.Context, telegramID int64) (*models.User, error)
	GetByID(ctx context.Context, id uuid.UUID) (*models.User, error)
}

// hugAcceptor is the interface for accepting/declining hugs from Telegram.
type hugAcceptor interface {
	AcceptHug(ctx context.Context, hugID, receiverID uuid.UUID) (*models.Hug, error)
	DeclineHug(ctx context.Context, hugID, receiverID uuid.UUID) error
}

// Bot is a long-polling Telegram bot that handles /start deep-link commands
// and inline keyboard callbacks for hug actions.
type Bot struct {
	tg        *tgbot.Bot
	client    *Client
	linkStore *LinkStore
	userRepo  botUserRepo
	hugSvc    hugAcceptor
	logger    *slog.Logger
	enabled   bool
}

// NewBot creates a new Telegram bot. If the client is disabled (no token),
// Run() will be a no-op and SendHugSuggestion will fall back to plain messages.
func NewBot(client *Client, linkStore *LinkStore, userRepo botUserRepo, hugSvc hugAcceptor, logger *slog.Logger) *Bot {
	b := &Bot{
		client:    client,
		linkStore: linkStore,
		userRepo:  userRepo,
		hugSvc:    hugSvc,
		logger:    logger,
	}

	if !client.Enabled() {
		return b
	}

	tg, err := tgbot.New(client.token, tgbot.WithDefaultHandler(b.handleUpdate))
	if err != nil {
		logger.Error("telegram bot: failed to create", "error", err)
		return b
	}

	b.tg = tg
	b.enabled = true
	return b
}

// Run starts the long-polling bot. Blocks until ctx is cancelled.
func (b *Bot) Run(ctx context.Context) {
	if !b.enabled {
		b.logger.Info("telegram bot disabled (no token)")
		return
	}

	b.logger.Info("telegram bot started (long-polling)")
	b.tg.Start(ctx)
	b.logger.Info("telegram bot stopped")
}

// SendHugSuggestion sends a hug suggestion notification with Accept/Decline buttons.
func (b *Bot) SendHugSuggestion(ctx context.Context, receiverID uuid.UUID, hugID uuid.UUID, giverName string) {
	telegramID, err := b.userRepo.GetTelegramID(ctx, receiverID)
	if err != nil || telegramID == nil {
		return
	}

	text := fmt.Sprintf("🤗 <b>%s</b> хочет тебя обнять!", tgbot.EscapeMarkdownUnescaped(giverName))

	if !b.enabled {
		// Fallback to plain message via raw client
		_ = b.client.SendMessage(*telegramID, text)
		return
	}

	keyboard := &tgmodels.InlineKeyboardMarkup{
		InlineKeyboard: [][]tgmodels.InlineKeyboardButton{
			{
				{Text: "Обнять 🤗", CallbackData: "accept:" + hugID.String()},
				{Text: "Отклонить", CallbackData: "decline:" + hugID.String()},
			},
		},
	}

	_, err = b.tg.SendMessage(ctx, &tgbot.SendMessageParams{
		ChatID:      *telegramID,
		Text:        text,
		ParseMode:   tgmodels.ParseModeHTML,
		ReplyMarkup: keyboard,
	})
	if err != nil {
		b.logger.Error("telegram bot: failed to send hug suggestion", "receiver_id", receiverID, "error", err)
	}
}

func (b *Bot) handleUpdate(ctx context.Context, bot *tgbot.Bot, update *tgmodels.Update) {
	if update.CallbackQuery != nil {
		b.handleCallback(ctx, bot, update.CallbackQuery)
		return
	}

	if update.Message != nil && strings.HasPrefix(update.Message.Text, "/start") {
		b.handleStart(ctx, bot, update.Message)
	}
}

func (b *Bot) handleStart(ctx context.Context, bot *tgbot.Bot, msg *tgmodels.Message) {
	chatID := msg.Chat.ID

	parts := strings.SplitN(msg.Text, " ", 2)
	if len(parts) < 2 || strings.TrimSpace(parts[1]) == "" {
		b.reply(ctx, bot, chatID, "Чтобы привязать аккаунт, используй настройки на сайте")
		return
	}

	token := strings.TrimSpace(parts[1])
	userID, ok := b.linkStore.ConsumeToken(token)
	if !ok {
		b.reply(ctx, bot, chatID, "Ссылка недействительна или истекла. Попробуй снова через настройки приложения")
		return
	}

	taken, err := b.userRepo.IsTelegramIDTaken(ctx, chatID, userID)
	if err != nil {
		b.logger.Error("telegram bot: failed to check telegram_id", "error", err)
		b.reply(ctx, bot, chatID, "Произошла ошибка. Попробуй позже :(")
		return
	}
	if taken {
		b.reply(ctx, bot, chatID, "Этот Telegram аккаунт уже привязан к другому пользователю :(")
		return
	}

	_, err = b.userRepo.SetTelegramID(ctx, userID, chatID)
	if err != nil {
		b.logger.Error("telegram bot: failed to set telegram_id", "user_id", userID, "chat_id", chatID, "error", err)
		b.reply(ctx, bot, chatID, "Произошла ошибка при привязке. Попробуй позже :(")
		return
	}

	b.logger.Info("telegram bot: account linked", "user_id", userID, "chat_id", chatID)
	b.reply(ctx, bot, chatID, "✅ Аккаунт привязан! Теперь ты не пропустишь обнимашки от любимых продовцев")
}

func (b *Bot) handleCallback(ctx context.Context, bot *tgbot.Bot, cb *tgmodels.CallbackQuery) {
	data := cb.Data
	chatID := cb.Message.Message.Chat.ID

	var action, hugIDStr string
	if strings.HasPrefix(data, "accept:") {
		action = "accept"
		hugIDStr = strings.TrimPrefix(data, "accept:")
	} else if strings.HasPrefix(data, "decline:") {
		action = "decline"
		hugIDStr = strings.TrimPrefix(data, "decline:")
	} else {
		return
	}

	hugID, err := uuid.Parse(hugIDStr)
	if err != nil {
		b.answerCallback(ctx, bot, cb.ID, "Ошибка: некорректные данные")
		return
	}

	// Look up the user by their Telegram chat ID
	user, err := b.userRepo.GetByTelegramID(ctx, chatID)
	if err != nil {
		b.answerCallback(ctx, bot, cb.ID, "Ваш Telegram не привязан к аккаунту")
		return
	}

	msgID := cb.Message.Message.ID
	originalText := cb.Message.Message.Text

	switch action {
	case "accept":
		_, err = b.hugSvc.AcceptHug(ctx, hugID, user.ID)
		if err != nil {
			b.logger.Error("telegram bot: failed to accept hug", "hug_id", hugID, "error", err)
			b.answerCallback(ctx, bot, cb.ID, "Не удалось принять объятие: "+friendlyError(err))
			return
		}
		b.answerCallback(ctx, bot, cb.ID, "Объятие принято! 🤗")
		b.editMessageText(ctx, bot, chatID, msgID, originalText+"\n\n✅ <b>Принято!</b>")

	case "decline":
		err = b.hugSvc.DeclineHug(ctx, hugID, user.ID)
		if err != nil {
			b.logger.Error("telegram bot: failed to decline hug", "hug_id", hugID, "error", err)
			b.answerCallback(ctx, bot, cb.ID, "Не удалось отклонить: "+friendlyError(err))
			return
		}
		b.answerCallback(ctx, bot, cb.ID, "Объятие отклонено")
		b.editMessageText(ctx, bot, chatID, msgID, originalText+"\n\n❌ <b>Отклонено</b>")
	}
}

func (b *Bot) reply(ctx context.Context, bot *tgbot.Bot, chatID int64, text string) {
	_, err := bot.SendMessage(ctx, &tgbot.SendMessageParams{
		ChatID: chatID,
		Text:   text,
	})
	if err != nil {
		b.logger.Error("telegram bot: failed to send message", "chat_id", chatID, "error", err)
	}
}

func (b *Bot) answerCallback(ctx context.Context, bot *tgbot.Bot, callbackID string, text string) {
	_, err := bot.AnswerCallbackQuery(ctx, &tgbot.AnswerCallbackQueryParams{
		CallbackQueryID: callbackID,
		Text:            text,
		ShowAlert:       false,
	})
	if err != nil {
		b.logger.Error("telegram bot: failed to answer callback", "error", err)
	}
}

func (b *Bot) editMessageText(ctx context.Context, bot *tgbot.Bot, chatID int64, messageID int, newText string) {
	_, err := bot.EditMessageText(ctx, &tgbot.EditMessageTextParams{
		ChatID:    chatID,
		MessageID: messageID,
		Text:      newText,
		ParseMode: tgmodels.ParseModeHTML,
	})
	if err != nil {
		b.logger.Error("telegram bot: failed to edit message", "error", err)
	}
}

func friendlyError(err error) string {
	msg := err.Error()
	switch {
	case strings.Contains(msg, "not found"):
		return "объятие не найдено"
	case strings.Contains(msg, "not pending"):
		return "объятие уже обработано"
	case strings.Contains(msg, "expired"):
		return "объятие истекло"
	default:
		return "попробуйте позже"
	}
}

// DeepLinkURL returns the t.me deep-link URL for a given token and bot username.
func DeepLinkURL(botUsername, token string) string {
	return fmt.Sprintf("https://t.me/%s?start=%s", botUsername, token)
}

File: backend/src/telegram/bot.rs
//! Long-polling Telegram bot — handles `/start <token>` deep-links for both
//! account-linking and Telegram-login flows, plus inline-button callbacks for
//! hug accept/decline from notification messages.

use std::sync::Arc;
use std::time::Duration;

use sqlx::PgPool;
use tracing::{error, info, warn};
use uuid::Uuid;

use crate::error::AppError;
use crate::repo;
use crate::service::{HugService, UserService};
use crate::telegram::client::{
    CallbackQuery, Client, InlineKeyboardButton, InlineKeyboardMarkup, Message, Update,
};
use crate::telegram::{LinkStore, LoginStore, TelegramUserInfo};

pub struct Bot {
    client: Arc<Client>,
    link_store: Arc<LinkStore>,
    login_store: Arc<LoginStore>,
    pool: PgPool,
    hug: Arc<HugService>,
    user: Arc<UserService>,
}

impl Bot {
    pub fn new(
        client: Arc<Client>,
        link_store: Arc<LinkStore>,
        login_store: Arc<LoginStore>,
        pool: PgPool,
        hug: Arc<HugService>,
        user: Arc<UserService>,
    ) -> Self {
        Self {
            client,
            link_store,
            login_store,
            pool,
            hug,
            user,
        }
    }

    pub fn client(&self) -> &Client {
        &self.client
    }

    /// Run long-polling until the task is aborted.
    pub async fn run(self: Arc<Self>) {
        if !self.client.enabled() {
            info!("telegram bot disabled (no token)");
            return;
        }
        info!("telegram bot started (long-polling)");
        let mut offset: i64 = 0;
        loop {
            match self.client.get_updates(offset, 30).await {
                Ok(updates) => {
                    for update in updates {
                        offset = update.update_id + 1;
                        let me = self.clone();
                        tokio::spawn(async move {
                            me.dispatch(update).await;
                        });
                    }
                }
                Err(err) => {
                    warn!(%err, "telegram bot: getUpdates failed");
                    tokio::time::sleep(Duration::from_secs(3)).await;
                }
            }
        }
    }

    pub async fn send_hug_suggestion(
        &self,
        receiver_id: Uuid,
        hug_id: Uuid,
        giver_name: &str,
        phrase: &str,
        comment: Option<&str>,
    ) {
        let telegram_id = match repo::user::get_telegram_id(&self.pool, receiver_id).await {
            Ok(Some(id)) => id,
            _ => return,
        };
        let mut text = format!("🤗 <b>{}</b> {}!", escape_html(giver_name), phrase);
        if let Some(c) = comment {
            if !c.is_empty() {
                text.push_str(&format!("\n\n💬 <i>{}</i>", escape_html(c)));
            }
        }
        let keyboard = InlineKeyboardMarkup {
            inline_keyboard: vec![vec![
                InlineKeyboardButton {
                    text: "Обнять 🤗".into(),
                    callback_data: Some(format!("accept:{hug_id}")),
                },
                InlineKeyboardButton {
                    text: "Отклонить".into(),
                    callback_data: Some(format!("decline:{hug_id}")),
                },
            ]],
        };
        if let Err(err) = self
            .client
            .send_message_with_kb(telegram_id, &text, &keyboard)
            .await
        {
            error!(%err, "telegram bot: send_hug_suggestion failed");
        }
    }

    async fn dispatch(self: Arc<Self>, update: Update) {
        if let Some(cb) = update.callback_query {
            self.handle_callback(cb).await;
            return;
        }
        if let Some(msg) = update.message {
            if msg.text.as_deref().map(|t| t.starts_with("/start")).unwrap_or(false) {
                self.handle_start(msg).await;
            }
        }
    }

    async fn handle_start(&self, msg: Message) {
        let chat_id = msg.chat.id;
        let text = msg.text.clone().unwrap_or_default();
        let parts: Vec<&str> = text.splitn(2, ' ').collect();
        if parts.len() < 2 || parts[1].trim().is_empty() {
            self.reply(chat_id, "Чтобы привязать аккаунт, используй настройки на сайте").await;
            return;
        }
        let token = parts[1].trim();

        if let Some(rest) = token.strip_prefix("login_") {
            self.handle_login_start(&msg, rest).await;
            return;
        }

        let user_id = match self.link_store.consume_token(token) {
            Some(id) => id,
            None => {
                self.reply(chat_id, "Ссылка недействительна или истекла. Попробуй снова через настройки приложения").await;
                return;
            }
        };

        let taken = match repo::user::is_telegram_id_taken(&self.pool, chat_id, user_id).await {
            Ok(t) => t,
            Err(err) => {
                error!(?err, "telegram bot: failed to check telegram_id");
                self.reply(chat_id, "Произошла ошибка. Попробуй позже :(").await;
                return;
            }
        };
        if taken {
            self.reply(chat_id, "Этот Telegram аккаунт уже привязан к другому пользователю :(").await;
            return;
        }

        if let Err(err) = repo::user::set_telegram_id(&self.pool, user_id, chat_id).await {
            error!(?err, %user_id, %chat_id, "telegram bot: failed to set telegram_id");
            self.reply(chat_id, "Произошла ошибка при привязке. Попробуй позже :(").await;
            return;
        }
        info!(%user_id, %chat_id, "telegram bot: account linked");
        self.reply(chat_id, "✅ Аккаунт привязан! Теперь ты не пропустишь обнимашки от любимых продовцев").await;
    }

    async fn handle_login_start(&self, msg: &Message, bot_token: &str) {
        let chat_id = msg.chat.id;
        let from = match &msg.from {
            Some(u) => u.clone(),
            None => {
                self.reply(chat_id, "Не удалось определить отправителя").await;
                return;
            }
        };

        let poll_token = match self.login_store.consume_bot_token(bot_token) {
            Some(p) => p,
            None => {
                self.reply(chat_id, "Ссылка недействительна или истекла. Попробуй снова").await;
                return;
            }
        };

        let info = TelegramUserInfo {
            telegram_id: chat_id,
            username: from.username.unwrap_or_default(),
            first_name: from.first_name,
            last_name: from.last_name.unwrap_or_default(),
        };

        match self.user.login_via_telegram(&info).await {
            Ok(user) => {
                self.login_store.authenticate(&poll_token, user.id);
                info!(%user.id, %chat_id, "telegram bot: login successful");
                self.reply(chat_id, "✅ Вход выполнен! Можешь вернуться в приложение").await;
            }
            Err(err) => {
                let friendly = friendly_login_error(&err);
                self.login_store
                    .fail(&poll_token, friendly.clone());
                error!(?err, %chat_id, "telegram bot: login failed");
                self.reply(chat_id, &format!("Не удалось войти: {friendly}")).await;
            }
        }
    }

    async fn handle_callback(&self, cb: CallbackQuery) {
        let data = match &cb.data {
            Some(d) => d.clone(),
            None => return,
        };
        let chat_id = cb
            .message
            .as_ref()
            .map(|m| m.chat.id)
            .unwrap_or(cb.from.id);
        let (action, id_str) = if let Some(rest) = data.strip_prefix("accept:") {
            ("accept", rest)
        } else if let Some(rest) = data.strip_prefix("decline:") {
            ("decline", rest)
        } else {
            return;
        };
        let Ok(hug_id) = Uuid::parse_str(id_str) else {
            self.answer_callback(&cb.id, "Ошибка: некорректные данные").await;
            return;
        };

        let user = match repo::user::get_by_telegram_id(&self.pool, chat_id).await {
            Ok(u) => u,
            Err(_) => {
                self.answer_callback(&cb.id, "Ваш Telegram не привязан к аккаунту").await;
                return;
            }
        };

        let original_text = cb.message.as_ref().and_then(|m| m.text.clone()).unwrap_or_default();
        let msg_id = cb.message.as_ref().map(|m| m.message_id).unwrap_or(0);

        match action {
            "accept" => match self.hug.accept_hug(hug_id, user.id).await {
                Ok(_) => {
                    self.answer_callback(&cb.id, "Объятие принято! 🤗").await;
                    let new_text = format!("{}\n\n✅ <b>Принято!</b>", escape_html(&original_text));
                    let _ = self
                        .client
                        .edit_message_text_html(chat_id, msg_id, &new_text)
                        .await;
                }
                Err(err) => {
                    error!(?err, %hug_id, "telegram bot: accept failed");
                    self.answer_callback(&cb.id, &format!("Не удалось принять: {}", friendly_hug_error(&err))).await;
                }
            },
            "decline" => match self.hug.decline_hug(hug_id, user.id).await {
                Ok(_) => {
                    self.answer_callback(&cb.id, "Объятие отклонено").await;
                    let new_text =
                        format!("{}\n\n❌ <b>Отклонено</b>", escape_html(&original_text));
                    let _ = self
                        .client
                        .edit_message_text_html(chat_id, msg_id, &new_text)
                        .await;
                }
                Err(err) => {
                    error!(?err, %hug_id, "telegram bot: decline failed");
                    self.answer_callback(&cb.id, &format!("Не удалось отклонить: {}", friendly_hug_error(&err))).await;
                }
            },
            _ => {}
        }
    }

    async fn reply(&self, chat_id: i64, text: &str) {
        if let Err(err) = self.client.send_message_html(chat_id, text).await {
            error!(%err, %chat_id, "telegram bot: failed to send message");
        }
    }

    async fn answer_callback(&self, callback_id: &str, text: &str) {
        if let Err(err) = self.client.answer_callback(callback_id, text).await {
            error!(%err, "telegram bot: failed to answer callback");
        }
    }
}

fn friendly_hug_error(err: &AppError) -> String {
    match err {
        AppError::HugNotFound => "объятие не найдено".into(),
        AppError::HugNotPending => "объятие уже обработано".into(),
        AppError::HugExpired => "объятие истекло".into(),
        _ => "попробуйте позже".into(),
    }
}

fn friendly_login_error(err: &AppError) -> String {
    match err {
        AppError::UserBanned => "ваш аккаунт заблокирован".into(),
        _ => "попробуйте позже".into(),
    }
}

fn escape_html(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
}
